### DISCUSSION TITLE
Can we launch plugins independently

### DISCUSSION THREAD

**Discussion by jjrrw174** [koreader/koreader#10805] [General]
Anyway to launch plugins independently? Specifically the reading stats one?

I have a PocketBook, and have tried playing around with the launch script. Wondering if someone has any ideas.

Thanks

**Comment by Frenzie** [comment]
Perhaps see https://github.com/koreader/koreader/blob/0db7509fe244c7a5ad6710a8aea8b9d22143f9c9/tools/wbuilder.lua but it might help to explain a bit more about what you're looking for.

**Reply by jjrrw174** [reply]
Thanks I'll take a look.

I'm trying to create a pocketbook .app file (sh script is what I want but they can be .c files not familiar with c though) so I can directly launch into statistics

**Reply by NiLuJe** [reply]
Do you actually want *just* the statistics, or do you want to launch a fully usable KOReader instance that just happens to start on a specific  Statistics view?

**Reply by jjrrw174** [reply]
I'm assuming by just statistics you mean the SQL behinde the statistics plugin. Please, correct me if I assumed wrong

I would like to launch the entire code behinde the statistics. The calender option, day option etc. 

I could extract the SQL or make my own export but I'm not familiar with C and I'm not familiar with making GUI apps. I'm more of security IT person :)

Thanks!

**Reply by NiLuJe** [reply]
I meant the whole existing Statistics UI/thing, but in a vacuum of not-really-KOReader (e.g., anything that would involve actual KOReader base features would not work and/or crash/exit ;p).

This is what the original request sounded like to me, and I wanted to highlight the fact that it felt weird ;).

Just aking a proper KOReader instance to launch whatever Stats view you want sounds much saner/easier.

**Reply by jjrrw174** [reply]
Let's go with sane then haha 😄 

Any hints or resources? I took a look at the pocketbook start script and it looks like it's the iv2sh that creates the process... I think...

If I'm going down the wrong path let me know

**Reply by NiLuJe** [reply]
You'll want to look at how reader.lua handles the various existing start with options, and basically going with the standard FM one (it's probably going to be faster), then tacking on whatever the Statistics menu entry of your choice actually does ;).

(You need either a Reader or a FileManager for plugins to exist, so that's probably going to be easier this way than trying to replicate the slightly fancy way plugins are handled in a vacuum).

**Reply by Frenzie** [reply]
> then tacking on whatever the Statistics menu entry of your choice actually does ;).

Not necessarily; what the Gesture Manager does is probably more on point. Which reminds me you can probably already do this and if you can't it may classify as a bug: create a profile and set it to autostart.

In this context, you'd have an FR to be able to pass an argument to autostart a profile/action or something like that. If I ping @hius07 I suspect we just might have it tomorrow given how enthusiastic he's been lately. :partying_face:

**Reply by poire-z** [reply]
> have an FR to be able to pass an argument to autostart a profile/action or something like that

We'd probably better leave all this (argument parsing, or less intrusive: environment variable presence check) to a user patch.

**Reply by Frenzie** [reply]
Good point. But in any case most of the functionality is there; it's basically just a profile with autostart, except you want to make it conditional based on a command line argument or something.

**Reply by NiLuJe** [reply]
I'm not a profile user, so I'm wholly unfamiliar with all that this implies ;).

**Reply by jjrrw174** [reply]
Likewise haha. I'll look at it though

**Reply by poire-z** [reply]
See https://github.com/koreader/koreader/wiki/User-patches

[2-start-with-stats.lua.txt](https://github.com/koreader/koreader/files/12336482/2-start-with-stats.lua.txt) (remove the .txt extension, drop that in koreader/patches/

```lua
if not os.getenv("KO_START_WITH_STATS") then return end

-- Be sure we don't save our hacked settings
G_reader_settings.flush = function() end
-- Force start with filemanager
G_reader_settings:delSetting("start_with")

local UIManager = require("ui/uimanager")
local FileManager = require("apps/filemanager/filemanager")

FileManager_init_orig = FileManager.init
FileManager.init = function(self)
    FileManager_init_orig(self)
    UIManager:nextTick(function()
        if os.getenv("KO_START_WITH_STATS") == "day" then
            self.statistics:onShowCalendarDayView()
        else
            self.statistics:onShowCalendarView()
        end
        -- We don't need you anymore FileManager!
        FileManager.instance:onClose()
    end)
end
```
(Seems to work just fine, and exit properly when done with statistics.)

Make one or two shell script wrappers that launch KOReader while previously setting the `KO_START_WITH_STATS` environment variable (dunno how it is launched on Pocketbook, so can't help more).
For me:
```
KO_START_WITH_STATS=day koreader-emu
KO_START_WITH_STATS=month koreader-emu
```

**Reply by jjrrw174** [reply]
Looks cool thanks I'll give it a shot :)

**Reply by jjrrw174** [reply]
@poire-z

I am using your user patch above and it does work as a separate application. I'm having trouble launching the default view with all the stats. I tried init() and onShowBookStats() but nothing happens.

Any tips on which function should be called?

**Reply by poire-z** [reply]
Not sure what you call "default view with all the stats". But as it's launched "as a separate application" (or "starting with File browser" for us :), there is no current book loaded, so the "Current book" menu is disabled in KOReader. So, it's probably normal that :onShowBookStats() is having issues in that unexpected situation.
My tip: give up :) Use RealKOReader:tm:

**Reply by jjrrw174** [reply]
I'm talking about this one. But now I'm seeing it display current book which I thought was over all lol my bad. Long week :)![Screenshot_20230819_095702_Firefox.jpg](https://github.com/koreader/koreader/assets/125767458/693616cb-bfdf-4a7e-8b9f-cb1d268f5cf9)

**Comment by hius07** [comment]
Sure, you can start KOReader with a stats page.
-set Start with - file browser
-create a profile with General - Statistics calendar view action
-check profile Autostart checkbox
But on closing the stats page KOReader remains working, not exited.

**Reply by ghost** [reply]
Thanks @hius07: This use case was exactly what I had in mind when doing https://github.com/koreader/koreader/pull/8049. See some screenshots there.

(My intention was to start the calculator on KOReader start.)

**Reply by ghost** [reply]
@poire-z I would delete the user-patch-available label here, as with the autostart functionality we have a smarter solution. ;)
